backend/scoutfs/scoutfs_compat.go

```diff
@@ -58,147 +49,31 @@ func New(rootdir string, opts ScoutfsOpts) (*ScoutFS, error) {
-	// later to link file into namespace
-	f := os.NewFile(uintptr(fd), filepath.Join(procfddir, strconv.Itoa(fd)))
-
-	tmp := &tmpfile{
-		f:          f,
-		bucket:     bucket,
-		objname:    obj,
-		size:       size,
-		needsChown: doChown,
-		uid:        uid,
-		gid:        gid,
-		newDirPerm: s.newDirPerm,
-	}
-
-	if doChown {
-		err := f.Chown(uid, gid)
-		if err != nil {
-			return nil, fmt.Errorf("set temp file ownership: %w", err)
-		}
-	}
-
-	return tmp, nil
-}
-
-func (tmp *tmpfile) link() error {
-	// We use Linkat/Rename as the atomic operation for object puts. The
-	// upload is written to a temp (or unnamed/O_TMPFILE) file to not conflict
-	// with any other simultaneous uploads. The final operation is to move the
-	// temp file into place for the object. This ensures the object semantics
-	// of last upload completed wins and is not some combination of writes
-	// from simultaneous uploads.
-	objPath := filepath.Join(tmp.bucket, tmp.objname)
-	err := os.Remove(objPath)
-	if err != nil && !errors.Is(err, fs.ErrNotExist) {
-		return fmt.Errorf("remove stale path: %w", err)
-	}
-
-	dir := filepath.Dir(objPath)
-
-	err = backend.MkdirAll(dir, tmp.uid, tmp.gid, tmp.needsChown, tmp.newDirPerm)
-	if err != nil {
-		return fmt.Errorf("make parent dir: %w", err)
-	}
-
-	procdir, err := os.Open(procfddir)
+func moveData(from *os.File, to *os.File) error {
+	// May fail if the files are not 4K aligned; check for alignment
+	ffi, err := from.Stat()
 	if err != nil {
-		return fmt.Errorf("open proc dir: %w", err)
+		return fmt.Errorf("stat from: %v", err)
 	}
-	defer procdir.Close()
-
-	dirf, err := os.Open(dir)
+	tfi, err := to.Stat()
 	if err != nil {
-		return fmt.Errorf("open parent dir: %w", err)
+		return fmt.Errorf("stat to: %v", err)
 	}
-	defer dirf.Close()
-
-	for {
-		err = unix.Linkat(int(procdir.Fd()), filepath.Base(tmp.f.Name()),
-			int(dirf.Fd()), filepath.Base(objPath), unix.AT_SYMLINK_FOLLOW)
-		if errors.Is(err, fs.ErrExist) {
-			err := os.Remove(objPath)
-			if err != nil && !errors.Is(err, fs.ErrNotExist) {
-				return fmt.Errorf("remove stale path: %w", err)
-			}
-			continue
-		}
-		if err != nil {
-			return fmt.Errorf("link tmpfile: %w", err)
-		}
-		break
+	if ffi.Size()%4096 != 0 || tfi.Size()%4096 != 0 {
+		return os.ErrInvalid
 	}
 
-	err = tmp.f.Close()
+	//
+	err = scoutfs.MoveData(from, to)
 	if err != nil {
-		return fmt.Errorf("close tmpfile: %w", err)
+		debuglogger.Logf("ScoutFs MoveData failed: %v", err)
 	}
-
-	return nil
-}
-
-func (tmp *tmpfile) Write(b []byte) (int, error) {
-	if int64(len(b)) > tmp.size {
-		return 0, fmt.Errorf("write exceeds content length %v", tmp.size)
-	}
-
-	n, err := tmp.f.Write(b)
-	tmp.size -= int64(n)
-	return n, err
-}
-
-func (tmp *tmpfile) cleanup() {
-	tmp.f.Close()
-}
-
-func (tmp *tmpfile) File() *os.File {
-	return tmp.f
-}
-
-func moveData(from *os.File, to *os.File) error {
-	return scoutfs.MoveData(from, to)
+	return err
 }
```

Comment:
its important that we call scoutfs.MoveData() for combining the final mulitpart object in scoutfs because scoutfs doesnt support a metadata only copy_file_range(). So without this, this will double the I/O to the filesystem for multipart uploads. I think we can still use scoutfs.MoveData() with a mktemp file instead of using O_TMPFILE to have a fix that will solve the issue here.

Reply:
oh! yes, sorry I missed that it was passing that function to the posix complete mp upload.